### DISCUSSION TITLE
Setting Up Union Type With Multiple Entities

### DISCUSSION THREAD

**Discussion by cameronmakarchuk** [hayes/pothos#1198] [Q&A] [answered]
Hi, just wondering the best approach for this as I'm running into a "property not found" error with my setup.

I have 3 subgraphs (Products, Content, Listing). In Content I have a union type that includes 2 entities (ProductPayload and ListingPayload):

```
UnionType = Builder.unionType('UnionType', {
   types: [ProductType, ListingType, AppType],
   resolveType: ({contentType}) => {
       ... if (contentType === 'PRODUCT') {
          return ProductType
      }

      ..... other conditions...
   }
})
```

In the `resolveType()` I'm getting an error saying "contentType does not exist on ...." because the keys for ProductPayload and Listing entities don't include `contentType`.

```
const SampleReference= builder.objectRef<SampleBackingType>(SampleReference').implement({
	fields: (t) => ({
		id: t.exposeString('id'),
		label: t.exposeString('label', { nullable: true }),
		contentType: t.field({
			type: CONSTANTS.ReferenceContentType,
			resolve: ({ contentType }) => {
				return contentType;
			},
		}),
		content: t.field({
			type: BlockReferenceContent,
			nullable: true,
			resolve: ({ contentType, id, ...additionalFields }, _, __, { variableValues }) => {
				if (!id || !contentType) {
					return null;
				}

				const { language, channel, client } = variableValues;

				let result = null;
				switch (contentType) {
					case CONSTANTS.ReferenceContentTypeEnum.PRODUCT:
						result = {
							id,
							channel,
							language,
							client,
						};
						break;
					case CONSTANTS.ReferenceContentTypeEnum.COLLECTION:
						result = {
							id,
							channel,
							language,
							client,
						};
						break;
					case CONSTANTS.ReferenceContentTypeEnum.APP_URL:
						result = {
							id,
							...additionalFields,
						};
						break;
					default:
						result = null;
						break;
				}

				if (!result) {
					return null;
				}

				return {
					...result,
					contentType,
				};
			},
		}),
	}),
});
```

I thought adding `contentType` to the return at the bottom would solve the issue but it didn't.

Any suggestions on where I might have the setup wrong?

**Comment by hayes** [comment, marked as answer]
There are a number of different options here.

In your resolve type, you can do something like `resolveType: (source) => if ('contentType' in source && source.contentType === 'PRODUCT')

Adding contentType to the returned values in the parent resolver isn't suffient because typescript doesn't know about all the fields that return your union type, it only knows about the types that make up the union type, so you would need to add `contentType` to all of the individual types that make up the union.

Another option is to remove the need for the resolveType function.  You can do this by adding `__typename` to the returned values instead.  Alternatively, you can use the `brandWithType` utility function

**Comment by cameronmakarchuk** [comment]
Thanks so much for the quick reply! That first option solved the issue, but I may play around with the other options too to see which fits best.
